crypto/ec/ecp_nistp384.c

```diff
@@ -0,0 +1,1975 @@
+           + ((uint128_t) in1[1]) * in2[4]
+           + ((uint128_t) in1[2]) * in2[3]
+           + ((uint128_t) in1[3]) * in2[2]
+           + ((uint128_t) in1[4]) * in2[1]
+           + ((uint128_t) in1[5]) * in2[0];
+
+    out[6] = ((uint128_t) in1[0]) * in2[6]
+           + ((uint128_t) in1[1]) * in2[5]
+           + ((uint128_t) in1[2]) * in2[4]
+           + ((uint128_t) in1[3]) * in2[3]
+           + ((uint128_t) in1[4]) * in2[2]
+           + ((uint128_t) in1[5]) * in2[1]
+           + ((uint128_t) in1[6]) * in2[0];
+
+    out[7] = ((uint128_t) in1[1]) * in2[6]
+           + ((uint128_t) in1[2]) * in2[5]
+           + ((uint128_t) in1[3]) * in2[4]
+           + ((uint128_t) in1[4]) * in2[3]
+           + ((uint128_t) in1[5]) * in2[2]
+           + ((uint128_t) in1[6]) * in2[1];
+
+    out[8] = ((uint128_t) in1[2]) * in2[6]
+           + ((uint128_t) in1[3]) * in2[5]
+           + ((uint128_t) in1[4]) * in2[4]
+           + ((uint128_t) in1[5]) * in2[3]
+           + ((uint128_t) in1[6]) * in2[2];
+
+    out[9] = ((uint128_t) in1[3]) * in2[6]
+           + ((uint128_t) in1[4]) * in2[5]
+           + ((uint128_t) in1[5]) * in2[4]
+           + ((uint128_t) in1[6]) * in2[3];
+
+    out[10] = ((uint128_t) in1[4]) * in2[6]
+            + ((uint128_t) in1[5]) * in2[5]
+            + ((uint128_t) in1[6]) * in2[4];
+
+    out[11] = ((uint128_t) in1[5]) * in2[6]
+            + ((uint128_t) in1[6]) * in2[5];
+
+    out[12] = ((uint128_t) in1[6]) * in2[6];
+}
+
+/*-
+ * Reduce thirteen 128-bit coefficients to seven 64-bit coefficients.
+ * in[i] < 2^128 - 2^125
+ * out[i] < 2^56 for i < 6,
+ * out[6] <= 2^48
+ */
+static void felem_reduce(felem out, const widefelem in)
+{
+    /*
+     * In order to prevent underflow, we add a multiple of p before subtracting.
+     * Use telescopic sums to represent 2^76 * p redundantly with each limb
+     * of the form 2^124 + ...
+     */
+    static const widelimb two124m68 = (((widelimb) 1) << 124)
+                                    - (((widelimb) 1) << 68);
+    static const widelimb two124m116m68 = (((widelimb) 1) << 124)
+                                        - (((widelimb) 1) << 116)
+                                        - (((widelimb) 1) << 68);
+    static const widelimb two124p108m76 = (((widelimb) 1) << 124)
+                                        + (((widelimb) 1) << 108)
+                                        - (((widelimb) 1) << 76);
+    static const widelimb two124m92m68 = (((widelimb) 1) << 124)
+                                       - (((widelimb) 1) << 92)
+                                       - (((widelimb) 1) << 68);
+    widelimb temp, output[9];
+    unsigned int i;
+
+    memcpy(output, in, sizeof(widelimb) * 9);
+
+    output[0] += two124p108m76;
+    output[1] += two124m116m68;
+    output[2] += two124m92m68;
+    output[3] += two124m68;
+    output[4] += two124m68;
+    output[5] += two124m68;
+    output[6] += two124m68;
+
+    /* eliminate in[6], ..., in[12] */
```

Comment:
Can you expand this comment to explain what the code below is doing.. I get the carry part but not this part..

Reply:
I'm deciding the most succinct way to comment this in code, but the underlying principle is this.
Let $p_{384} = 2^{384} - \delta$, then it follows that $2^{384} \equiv \delta \mod{p_{384}}$.

Thus for any suitable choice of $X$ and $Y$ such that our input is equal (or congruent) to $X + 2^{384}Y$ [1] which is in turn congruent to $X + \delta Y$.

The way you start off eliminating the high limbs (it should read 'eliminate in[7], ...., in[12]` is precisely with this substitution.

```c
    /* X += 2^128 Y */
    output[8] += in[12] >> 32;
    output[7] += (in[12] & 0xffffffff) << 24;

    /* X += 2^96 Y */
    output[7] += in[12] >> 8;
    output[6] += (in[12] & 0xff) << 48;
    
    /* X += (-2^32 + 1) Y */
    output[6] -= in[12] >> 16;
    output[5] -= ((in[12] & 0xffff) << 40);
    output[6] += in[12] >> 48;
    output[5] += (in[12] & 0xffffffffffff) << 8;
```

The reason why at each limb, and for term of $\delta = 2^{128} + 2^{96} + (-2^{32} + 1)$ there are multiple operations being performed is simply to reduce the maximum size of operands to prevent under/overflow.

in reality we do at least three such substitutions in this code. Firstly we use limbs `in[12]`, `in[11]`, `in[10]`, `in[9]` to be our $Y$. Then afterwards we take `output[8]`, `output[7]` as our $Y$ (the result of the previous substitution accumulated into these), eliminating these two limbs. Finally, we let `temp` be our choice of $Y$ (each of these with suitable scaling so that [1] holds) and substitute it (after partially carrying into `temp` in order to minimise the size of the resulting, reduced number). We perform one last final carry and then obtain the desired bound.